examples/training/resnet50/main.cpp

```diff
@@ -0,0 +1,238 @@
+  /// The database contains 10000 images.
+  /// Size: (1 + (32 * 32 * 3)) * 10000 = 30730000.
+  static const float NormMean[] = {0.485, 0.456, 0.406};
+  static const float NormStd[] = {0.229, 0.224, 0.225};
+
+  static const std::pair<float, float> Range = {0., 1.};
+
+  llvm::outs() << "Assigned Images/Labels bytes.\n";
+  auto LabelsH = Labels.getHandle<float>();
+  auto ImagesH = Images.getHandle<>();
+  float Scale = ((Range.second - Range.first) / 255.0);
+  float Bias = Range.first;
+
+  /// CIFAR image dimensions are 32x32, ResNet50 expects 224x224.
+  /// Simple scaling would be one CIFAR pixel copied into 7x7 area
+  constexpr unsigned SCALE = 7;
+  for (unsigned N = 0; N < CIFAR_NUM_IMAGES; ++N) {
+    unsigned long L = dbInput.get();
+    LabelsH.at({N, L}) = 1;
+    // Even ResNet50 model got trained in NCHW format
+    for (unsigned C = 0; C < IMAGE_COLORS; ++C) {
+      auto BRGC = IMAGE_COLORS - 1 - C;
+      for (unsigned H = 0; H < 32; ++H) {
+        for (unsigned W = 0; W < 32; ++W) {
+          // ResNet BGR color space vs CIFAR RGB.
+          auto Val = static_cast<float>(static_cast<uint8_t>(dbInput.get()));
+          // Normalize and scale image.
+          Val = (Val * Scale - NormMean[C]) / NormStd[C] + Bias;
+          for (unsigned SH = 0; SH < SCALE; ++SH) {
+            for (unsigned SW = 0; SW < SCALE; ++SW) {
+              ImagesH.at({N, BRGC, SH + H * SCALE, SW + W * SCALE}) = Val;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  const char *TextualLabels[LABELS_SIZE] = {
+      "airplane", "automobile", "bird",  "cat",  "deer",
+      "dog",      "frog",       "horse", "ship", "truck"};
+
+  size_t SampleCounter = 0;
+  int ReportRate = 16;
+
+  llvm::Timer timer("Training", "Training");
+
+  llvm::outs() << "Training.\n";
+  for (size_t e = 0; e < TrainingEpochs; ++e) {
+    llvm::outs() << "Training - epoch #" << e << " from total "
+                 << TrainingEpochs << "\n";
+    // Bind the images tensor to the input array A, and the labels tensor
+    // to the softmax node SM.
+    // Train network on all examples.
+    timer.startTimer();
+    runBatch(EE, Bindings, ReportRate, SampleCounter, {A, E},
+             {&Images, &Labels});
+    timer.stopTimer();
+
+    llvm::outs() << "Evaluating.\n";
+    unsigned Score = 0;
+    unsigned Total = 0;
+
+    for (unsigned int i = 0; i < CIFAR_NUM_IMAGES / MiniBatchSize; ++i) {
+      Tensor Sample(ElemKind::FloatTy, BatchImagesDims);
+      Sample.copyConsecutiveSlices(&Images, MiniBatchSize * i);
+      updateInputPlaceholders(Bindings, {A}, {&Sample});
+      EE.run(Bindings);
+
+      for (unsigned int Batch = 0; Batch < MiniBatchSize; ++Batch) {
+        auto CT = LabelsH.extractSlice(MiniBatchSize * i + Batch);
+        auto CH = CT.getHandle<>();
+        auto CorrectVal = CH.raw(0);
+        size_t CorrectIdx = 0;
+        for (size_t c = 0; c < 10; ++c) {
```

Comment:
This code and the code below seem identical but operate on different tensors. Consider refactoring.